test/sqllogictest/select_target_replica.slt

```diff
@@ -0,0 +1,48 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+mode cockroach
+
+statement ok
+CREATE TABLE test (a TEXT, b TEXT)
+
+statement ok
+INSERT INTO test VALUES('a', 'b')
+
+statement ok
+CREATE CLUSTER test
+  REPLICA replica_a (SIZE '1'),
+  REPLICA replica_b (SIZE '2')
+
+statement ok
+SET cluster = test
+
+statement ok
+SET cluster_replica = replica_a
+
+# Verify that simple SELECTs work.
+
+query TT
+SELECT * FROM test
+----
+a b
+
+# Verify that SELECTs on introspection tables work.
+
+query I
+SELECT * FROM mz_materializations
+----
+
+# Verify that targeting an unknown replica fails.
+
+statement ok
+SET cluster_replica = unknown
+
+query error cluster replica 'test.unknown' does not exist
+SELECT * FROM test
```

Comment:
It would be nice to verify that the responses we get back actually come from the selected replica. I tried this, but it's a) slow and b) flaky in CI:

```
# Verify that we are always hitting the same replica.
# We check this that the number of peeks in mz_peek_durations (which is a
# replica-specific introspection source) increases by one with every query.
# We sleep between checks to give dataflow logging time to catch up.

query I
SELECT mz_internal.mz_sleep(1)
----
NULL

query I
SELECT sum(count) FROM mz_peek_durations
----
2

query I
SELECT mz_internal.mz_sleep(1)
----
NULL

query I
SELECT sum(count) FROM mz_peek_durations
----
3

query I
SELECT mz_internal.mz_sleep(1)
----
NULL

query I
SELECT sum(count) FROM mz_peek_durations
----
4

# Verify that we see no prior peeks when targeting another replica.

statement ok
SET cluster_replica = replica_b

query I
SELECT sum(count) FROM mz_peek_durations
----
NULL
```

Does anyone have a better idea?

Reply:
Looks like both issues are fixed by specifying a sufficiently small INTROSPECTION GRANULARITY when creating the cluster.